File: jscomp/others/bs_internalMutableAVLSet.ml


module N = Bs_internalAVLset

type 'elt node = 'elt N.node

type ('elt,'id) t0 = 'elt node Js.null

external unsafeCoerce : 'a Js.null -> 'a = "%identity"



let empty = N.empty0      
let isEmpty = N.isEmpty0
let singleton = N.singleton0
let min = N.min0
let max = N.max0
let iter = N.iter0      
let fold = N.fold0
let forAll = N.forAll0
let exists = N.exists0    
let filter = N.filter0
let partition = N.partition0
let cardinal = N.cardinal0
let elements = N.elements0 
let checkInvariant = N.checkInvariant
(* 
  L rotation, return root node
*)
let rotateWithLeftChild k2 = 
  let k1 = unsafeCoerce (N.left k2) in 
  N.(leftSet k2 (right k1)); 
  N.(rightSet k1 (return k2 ));
  let hlk2, hrk2 = N.(height (left k2), (height (right k2))) in  
  N.(hSet k2 
    (Pervasives.max hlk2 hrk2 + 1));
  let hlk1, hk2 = N.(height (left k1), (h k2)) in 
  N.(hSet k1 (Pervasives.max hlk1 hk2 + 1));
  k1  
(* right rotation *)
let rotateWithRightChild k1 =   
  let k2 = unsafeCoerce (N.right k1) in 
  N.(rightSet k1 (left k2));
  N.(leftSet k2 (return k1));
  let hlk1, hrk1 = N.((height (left k1)), (height (right k1))) in 
  N.(hSet k1 (Pervasives.max  hlk1 hrk1 + 1));
  let hrk2, hk1 = N.(height (right k2), (h k1)) in 
  N.(hSet k2 (Pervasives.max  hrk2 hk1 + 1));
  k2 

(*
  double l rotation
*)  
let doubleWithLeftChild k3 =   
  let v = rotateWithLeftChild (unsafeCoerce N.(left k3)) in 
  N.(leftSet k3 (return v ));
  rotateWithLeftChild k3 

let doubleWithRightChild k2 = 
  let v = rotateWithRightChild (unsafeCoerce N.(right k2)) in   
  N.(rightSet k2 (return v));
  rotateWithRightChild k2
type key = int 

let rec add (x : key) (t : _ t0) =   
  match N.toOpt t with 
  | None -> N.(return @@ node ~left:empty ~right:empty ~key:x ~h:1)
  | Some nt -> 
    let k = N.key nt in 
    if x = k then t 
    else
      begin 
        let l, r = N.(left nt, right nt) in 
        let t =
        (if x < k then             
           begin 
             N.leftSet nt (add x l);
             (if N.height l > 2 +  N.height r then 
                (if x < N.key (unsafeCoerce l) then 
                    rotateWithLeftChild nt 
                 else 
                    doubleWithLeftChild nt )
              else  nt )
           end
         else   
           begin 
             N.rightSet nt (add x r);
             (if N.height r > 2 + N.height l  then 
                (if N.key (unsafeCoerce r) < x then 
                   rotateWithRightChild nt 
                 else 
                   doubleWithRightChild nt
                ) else 
                nt
             )
           end
        ) in 
        let hlt, hrt = N.(height (left t),(height (right t))) in 
        N.hSet t 
          N.(Pervasives.max hlt hrt  + 1);
        N.return t
      end

File: lib/js/bs_internalMutableAVLSet.js
'use strict';

var Bs_internalAVLset = require("./bs_internalAVLset.js");
var Bs_internalSetInt = require("./bs_internalSetInt.js");

var N = 0;

var I = 0;

var empty = Bs_internalAVLset.empty0;

var isEmpty = Bs_internalAVLset.isEmpty0;

var singleton = Bs_internalAVLset.singleton0;

var min = Bs_internalAVLset.min0;

var max = Bs_internalAVLset.max0;

var iter = Bs_internalAVLset.iter0;

var fold = Bs_internalAVLset.fold0;

var forAll = Bs_internalAVLset.forAll0;

var exists = Bs_internalAVLset.exists0;

var filter = Bs_internalAVLset.filter0;

var partition = Bs_internalAVLset.partition0;

var length = Bs_internalAVLset.length0;

var elements = Bs_internalAVLset.elements0;

var toArray = Bs_internalAVLset.toArray0;

var checkInvariant = Bs_internalAVLset.checkInvariant;

var add = Bs_internalSetInt.addMutate;

var ofArray = Bs_internalSetInt.ofArray;

var cmp = Bs_internalSetInt.cmp;

var diff = Bs_internalSetInt.diff;

var eq = Bs_internalSetInt.eq;

var findOpt = Bs_internalSetInt.findOpt;

var split = Bs_internalSetInt.split;

var subset = Bs_internalSetInt.subset;

var inter = Bs_internalSetInt.inter;

var union = Bs_internalSetInt.union;

var remove = Bs_internalSetInt.remove;

var mem = Bs_internalSetInt.mem;

exports.N = N;
exports.I = I;
exports.empty = empty;
exports.isEmpty = isEmpty;
exports.singleton = singleton;
exports.min = min;
exports.max = max;
exports.iter = iter;
exports.fold = fold;
exports.forAll = forAll;
exports.exists = exists;
exports.filter = filter;
exports.partition = partition;
exports.length = length;
exports.elements = elements;
exports.toArray = toArray;
exports.checkInvariant = checkInvariant;
exports.add = add;
exports.ofArray = ofArray;
exports.cmp = cmp;
exports.diff = diff;
exports.eq = eq;
exports.findOpt = findOpt;
exports.split = split;
exports.subset = subset;
exports.inter = inter;
exports.union = union;
exports.remove = remove;
exports.mem = mem;
/* No side effect */
